### ISSUE TITLE
`**` glob does not match files in the root directory

### ISSUE THREAD

**Issue by zx8** [fish-shell/fish-shell#7222]
```
$ fish --version
fish, version 3.1.2-1083-ga30630a1d
```

---

### bash
```
$ mkdir foo
$ touch bar foo/bar
$ echo **/bar
bar foo/bar
```

### fish
```
$ mkdir foo
$ touch bar foo/bar
$ echo **/bar
foo/bar
```

Is this intentional or an oversight? The [documentation](https://fishshell.com/docs/current/#wildcards) says:

```
** matches any files and directories in the current directory and all of its subdirectories.
```

**Comment by gokulsoumya** [comment]
A *lone* `**` matches any files and directories in the current directory; the description you saw was an example. A few paragraphs above it is mentioned:

> `**` matches any string of characters. This includes matching an empty string. The matched string may include the / character; that is, it recurses into subdirectories.

And of particular interest in the specific example you mentioned is this line in the docs (emphasis mine):

> Note that augmenting this wildcard with other strings will not match files in the current working directory (`$PWD`) **if you separate the strings with a slash ("/")**.

So to match files in the current directory also, you specify the glob *without* the `/` like so:

```fish
echo **bar
```

**Comment by ridiculousfish** [comment]
Right, the forward slash is part of the glob and must be matched. I think fish shell is pretty committed to this decision. Closing, but thanks for filing!

**Comment by zx8** [comment]
> @gokulsoumya So to match files in the current directory also, you specify the glob _without_ the `/` like so:
> 
> ```fish
> $ echo **bar
> bar foo/bar
> ```

The only problem with this is that it will also match e.g. `foobar` in the current directory.

> @ridiculousfish Right, the forward slash is part of the glob and must be matched. I think fish shell is pretty committed to this decision. Closing, but thanks for filing!

That's fair enough, but it also makes sense to me if `**` matched `.` (i.e. the working directory) because then `**/bar` would pick up `./bar` and still be semantically correct, right?

**Comment by zx8** [comment]
Also, fish seems like the odd one out here:

### fish

```fish
$ echo **/bar
foo/bar
```

### bash

```sh
$ echo **/bar
bar foo/bar
```

### zsh

```sh
$ echo **/bar
bar foo/bar
```

### python

```python
Python 3.8.5 (default, Jul 21 2020, 10:48:26)
[Clang 11.0.3 (clang-1103.0.32.62)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import glob
>>> glob.glob("**/bar", recursive=True)
['bar', 'foo/bar']
```

### ruby

```ruby
$ irb
irb(main):001:0> Dir.glob('**/bar')
=> ["foo/bar", "bar"]
```

**Comment by ridiculousfish** [comment]
Ok, maybe we should change it.

**Comment by mqudsi** [comment]
I've been thrown by this before, tbh.

I think the current behavior is fishy (not in a good way), and since fish only uses globs for matching against files, it seems like it's not a stretch to incorporate path semantics into the glob matching.

That said, `git` behaves similar to fish:

```
mqudsi@Blitzkrieg /m/d/r/fish> git ls-files '*.in' | grep -v /
config_cmake.h.in
fish.pc.in
fish.spec.in
mqudsi@Blitzkrieg /m/d/r/fish> git ls-files '**/*.in' | grep -v /
```

**Comment by zx8** [comment]
@ridiculousfish Hope I'm not being too cheeky here, but notice your recent commuts to master are poking around the ** glob code - any chance you might be able to address this issue while you're in the thick of it? 🙏

**Comment by ridiculousfish** [comment]
I found a way to add this behavior without too much extra work. Most other glob implementations require that `**` appear in a segment by itself; fish does not which makes this trickier. I added a special case to incorporate that behavior; that is if `**` appears in a segment by itself we also match in the same directory as if that segment were missing. Updated docs too, tried to make them simpler.
